File: src/Functional/OCaml/example3.ml
module type Report = sig
  val generate : unit -> unit
end

module PDFReport : Report = struct
  let generate () = print_endline "Generating PDF report"
end

module HTMLReport : Report = struct
  let generate () = print_endline "Generating HTML report"
end

module type ReportFactory = sig
  val create_report : unit -> (module Report)
end

module PDFReportFactory : ReportFactory = struct
  let create_report () = (module PDFReport : Report)
end

module HTMLReportFactory : ReportFactory = struct
  let create_report () = (module HTMLReport : Report)
end

let use_factory (factory : (module ReportFactory)) =
  let module F = (val factory : ReportFactory) in
  let module R = (val F.create_report () : Report) in
  R.generate ()

(* Usage *)
let () =
  use_factory (module PDFReportFactory);
  use_factory (module HTMLReportFactory)

File: src/Functional/Scala/Example3.scala
trait Report {
  def generate(): Unit
}

class PDFReport extends Report {
  def generate(): Unit = println("Generating PDF report")
}

class HTMLReport extends Report {
  def generate(): Unit = println("Generating HTML report")
}

trait ReportFactory {
  def createReport(): Report
}

class PDFReportFactory extends ReportFactory {
  def createReport(): Report = new PDFReport
}

class HTMLReportFactory extends ReportFactory {
  def createReport(): Report = new HTMLReport
}

object Example3 {
  def useFactory(factory: ReportFactory): Unit = {
    val report = factory.createReport()
    report.generate()
  }

  def main(args: Array[String]): Unit = {
    useFactory(new PDFReportFactory)
    useFactory(new HTMLReportFactory)
  }
}
